### DISCUSSION TITLE
Is there a way to disable one lint on one line only?

### DISCUSSION THREAD

**Discussion by benesch** [microsoft/pyright#2088] [Q&A]
Suppose I have some code that necessarily violates Pyright's rules. For example:

```py
FOO = 'bar'
FOO = 'baz'
```

This redeclares a constant, which pyright flags because I have `reportConstantRedefinition=true`. I would like to be able to suppress this lint on this line only. As far as I can tell, that's not something I can do in pyright today? My options seem to be:

1. Add a `type: ignore` comment

    ```py
    FOO = 'bar'
    FOO = 'baz' # type: ignore
    ```

    This scopes the ignore to the line in question, but I don't want to ignore _any_ type error, just the constant redefinition error. If there's a different type error on that line, I want to know about it.

2. Add a `pyright: reportConstantRedefinition=false` comment to the file in question. This has the benefit of only ignoring constant redefinitions, but it seems to apply to the file as a whole. And I don't want to allow constant redefinitions _in general_, just on this one line.

In mypy I think you can say something like

```py
FOO = 'bar'
FOO = 'baz' # type: ignore [const-redefinition]
```

which ignores only the error specified in square brackets. Does pyright have anything similar that I've just missed? If not, would y'all be up for adding it?

**Comment by erictraut** [comment]
Pyright supports the PEP 484 standard type ignore comment of the form `# type: ignore`. It does not support square brackets afterward. This is not part of the standard, and diagnostic rule names are not consistent across type checkers. For those reasons, we have no plans to support this notation. In general, it is exceedingly rare to have more than one diagnostic on a single line anyway.

My personal philosophy is that `# type: ignore` should never be used in a code base. There are always better ways to handle an error — by disabling certain diagnostic checks globally if the code base doesn't conform to them, changing the code to conform to the checks you want to enable, adding missing type annotations, and (as a last resort) adding `cast` calls. My team manages a Python code base with nearly half a million lines of code, and most of it is strictly type checked (i.e. pyright's strict mode is enforced). We also use the `"enableTypeIgnoreComments"=false` setting to disallow `# type: ignore` comments.

**Reply by benesch** [reply]
Thanks for getting back to me.

> For those reasons, we have no plans to support this notation.

I certainly don't care about any particular notation. I'd be perfectly happy if pyright had its own syntax for this:

```
FOO = 'bar'
FOO = 'baz' # pyright: reportConstantRedefinition=not-this-line
```

> In general, it is exceedingly rare to have more than one diagnostic on a single line anyway.

So my motivations are not to ignore one diagnostic but not another; it's about making sure the opt-out doesn't hide *new* errors that might be introduced in the future. Concretely, suppose I have this code today:

```
def recompute_foo(x: int):
    return 'baz' + str(x)

FOO = 'bar'
FOO = recompute_foo(1) # type: ignore
```

In this case I am intentionally overriding the constant, and I'd like to tell pyright not to warn me about it. But suppose the type of `recompute_foo` changes in the future such that `1` is no longer a valid argument. That `type: ignore` comment means that type error will be suppressed.

> My personal philosophy is that `# type: ignore` should never be used in a code base.

I definitely agree with this part!

> There are always better ways to handle an error — by disabling certain diagnostic checks globally if the code base doesn't conform to them, changing the code to conform to the checks you want to enable, adding missing type annotations, and (as a last resort) adding `cast` calls.

But I'm not sure I agree that there is _always_ a better way to handle an error. I think I agree that for a general type usage violation, adding a `cast` call is a good last resort before reaching for `type: ignore`. But what's the equivalent for allowing constant redefinition juuust once? Being able to drop a comment that applies to just one constant redefinition violation feels morally equivalent to a call to `cast` to me.

**Reply by erictraut** [reply]
If you want to enforce that all-caps symbols are constant, then you can change your code to not violate that rule. Or you can decide that the file in question doesn't need to honor that rule (e.g. because it's a test module) and disable the rule for that file.
